app/categories/traversal.py

```diff
@@ -0,0 +1,212 @@
+            raise NotFound()
+
+        return route
+
+    def get_question_answer_map(self, path: str) -> list[dict[str, str]]:
+        """Create question-answer mapping using pre-calculated routes
+        Returns a list of dictionaries mapping questions and answers.
+        """
+        result: list[dict[str, str]] = []
+
+        if path == "":
+            return result
+        segments = path.split("/")
+
+        current_path = ""
+
+        for index, segment in enumerate(segments):
+            route: NavigationResult = self.route_cache.get(current_path)
+
+            if not route or not route.question_form:
+                break
+
+            if segment not in route.question_form.valid_choices():
+                result.append(
+                    {"question": route.question_form.title, "answer": "invalid choice"}
+                )
+                break
+
+            if not hasattr(route.question_form, "title"):
+                raise ValueError(f"{route.question_form} does not have a title")
+
+            answer_label = route.question_form.get_label(segment)
+            result.append(
+                {"question": route.question_form.title, "answer": answer_label}
+            )
+            current_path += segment if index == 0 else f"/{segment}"
+
+        return result
+
+    def get_all_valid_paths(self) -> list[str]:
+        """Return a sorted list of all pre-calculated valid paths"""
+        return sorted(list(self.route_cache.keys()))
+
+    def get_previous_page_url(self, path: str) -> str:
+        """Generate a URL for the previous question page.
+            This populates the back button for each question page.
+
+            If the previous page is a QuestionForm then previous_answer will be populated
+
+        Args:
+            path: Current URL path
+
+        Returns:
+            URL for the previous question page
+        """
+        path_segments = path.split("/")
+
+        if len(path_segments) < 2:
+            return url_for("categories.index")
+
+        previous_answer = path_segments.pop()
+        new_path = "/".join(path_segments)
+        previous_page = self.navigate_path(new_path)
+
+        return url_for(
+            "categories.question_page",
+            path=new_path,
+            previous_answer=previous_answer
+            if isinstance(previous_page, QuestionForm)
+            else None,
+        )
+
+    def get_all_user_journeys(self) -> dict[str, NavigationResult]:
+        """Generate dictionary of path components for the routing map debug page.
+        Only includes paths that end in a redirect."""
+        all_paths = self.get_all_valid_paths()
+
+        full_paths = {}
+
+        for path in all_paths:
+            result = self.route_cache[path]
+            if not result.is_redirect:
+                continue
+            full_paths[path] = result
+        return full_paths
+
+    def get_all_question_pages(self) -> dict[str, NavigationResult]:
```

Comment:
Should this be part of test code for now?

Reply:
Good point, I am not actually sure what is the best approach here as it is generally a useful function. However, I have moved it to the test code.